core/lib/src/local/form.rs

```diff
@@ -0,0 +1,222 @@
+        self
+    }
+
+    /// Adds a field for the file with the name `file_name`, content-type `ct`, and
+    /// file contents `data`.
+    pub fn file<N, V>(mut self, file_name: N, ct: ContentType, data: V) -> Self
+    where
+        N: Into<Option<&'v str>>,
+        V: AsRef<[u8]>,
+    {
+        self.0.push(LocalField::Data(
+            NameBuf::from("file"),
+            file_name.into(),
+            ct,
+            data.as_ref().into(),
+        ));
+        self
+    }
+
+    /// Add a data field with a content-type `ct` and binary `data`.
+    pub fn data<V>(mut self, ct: ContentType, data: V) -> Self
+    where
+        V: AsRef<[u8]>,
+    {
+        self.0.push(LocalField::Data(
+            NameBuf::from("file"),
+            None,
+            ct,
+            data.as_ref().into(),
+        ));
+        self
+    }
+
+    /// The full content-type for this form.
+    pub fn content_type(&self) -> ContentType {
+        if self.contains_data_field() {
+            return "multipart/form-data; boundary=X-BOUNDARY"
+                .parse::<ContentType>()
+                .unwrap();
+        }
+
+        ContentType::Form
+    }
+
+    /// The full body data for this form.
+    pub fn body_data(&self) -> Vec<u8> {
+        if self.contains_data_field() {
+            self.format_multipart()
+        } else {
+            self.format_simple()
+        }
+    }
+
+    fn contains_data_field(&self) -> bool {
+        self
+            .0
+            .iter()
+            .any(|field| matches!(field, LocalField::Data(..)))
+    }
+
+    fn format_simple(&self) -> Vec<u8> {
+        self.0.iter().fold(Vec::new(), |mut acc, field| {
+            match field {
+                LocalField::Value(name, value) => {
+                    acc.push(format!("{}={}", Uri::percent_encode(&format!("{}", name)), Uri::percent_encode(value)));
+                    return acc
+                },
+                _ => acc,
+            }
+        }).join("&").as_bytes().to_vec()
+    }
+
+    fn format_multipart(&self) -> Vec<u8> {
+        let mut body = self.0.iter().fold(Vec::new(), |mut acc, field| {
+            match field {
+                LocalField::Value(name, value) => {
+                    acc.push("--X-BOUNDARY".to_string());
+                    acc.push(format!("Content-Disposition: form-data; name=\"{}\"", name));
+                    acc.push("".to_string());
+                    acc.push(format!("{}", value));
+                },
+                LocalField::Data(name, file_name, content_type, data) => {
+                    acc.push("--X-BOUNDARY".to_string());
+                    acc.push(format!("Content-Disposition: form-data; name=\"{}\"; filename=\"{}\"", name, file_name.unwrap_or("")));
+                    acc.push(format!("Content-Type: {}", content_type));
+                    acc.push("".to_string());
+                    acc.push(format!("{}", String::from_utf8_lossy(data)));
```

Comment:
This is wrong; `data` may not be UTF-8, and this will incorrectly format the data. This shouldn't be a `String`, but part of the byte buffer.